src/orion/algo/registry.py

```diff
@@ -0,0 +1,147 @@
+        elif isinstance(trial_or_id, str):
+            trial_id = trial_or_id
+        else:
+            raise NotImplementedError(trial_or_id)
+        return trial_id in self._trials
+
+    def __getitem__(self, item: int | str) -> Trial:
+        if not isinstance(item, (str, int)):
+            raise KeyError(item)
+        if isinstance(item, str):
+            return self._trials[item]
+        trial_ids = list(self._trials.keys())
+        return self._trials[trial_ids[item]]
+
+    def __iter__(self) -> Iterator[Trial]:
+        return iter(self._trials.values())
+
+    def __len__(self) -> int:
+        return len(self._trials)
+
+    @property
+    def state_dict(self) -> dict:
+        return {"_trials": self._trials}
+
+    def set_state(self, statedict: dict) -> None:
+        self._trials = statedict["_trials"]
+
+    def has_suggested(self, trial: Trial) -> bool:
+        return _get_id(trial) in self
+
+    def has_observed(self, trial: Trial) -> bool:
+        trial_id = _get_id(trial)
+        if trial_id not in self:
+            return False
+        return self[trial_id].status in ("broken", "completed")
+
+    def register(self, trial: Trial) -> str:
+        """Register the given trial in the registry."""
+        trial_id = _get_id(trial)
+        if trial_id in self:
+            existing = self._trials[trial_id]
+            logger.debug(
+                f"Overwriting existing trial {existing} with new trial {trial}"
+            )
+        else:
+            logger.debug(
+                f"Registry {id(self)} Registering new trial {trial} ({len(self)} trials in total)"
+            )
+        trial_copy = copy.deepcopy(trial)
+        self._trials[trial_id] = trial_copy
+        return trial_id
+
+
+class RegistryMapping(Mapping[Trial, "list[Trial]"]):
+    """A map between the original and transformed registries.
+
+    This object is used in the `SpaceTransformAlgoWrapper` to check if a trial in the original space
+    has equivalent trials in the transformed space.
+
+    The goal is to make it so the algorithms don't have to care about the transforms/etc.
+    """
+
+    def __init__(self, original_registry: Registry, transformed_registry: Registry):
+        self.original_registry = original_registry
+        self.transformed_registry = transformed_registry
+        self._mapping: dict[str, set[str]] = defaultdict(set)
+
+    @property
+    def state_dict(self) -> dict:
+        return {
+            "original_registry": self.original_registry.state_dict,
+            "transformed_registry": self.transformed_registry.state_dict,
+            "_mapping": self._mapping,
+        }
+
+    def set_state(self, statedict: dict):
+        self.original_registry.set_state(statedict["original_registry"])
+        self.transformed_registry.set_state(statedict["transformed_registry"])
+        self._mapping = statedict["_mapping"]
+
+    def __iter__(self) -> Iterator[Trial]:
+        return iter(self.original_registry)
```

Comment:
Not sure why we would iterate on the mapping, but if we would do I would assume we want the mapped (original_trial, transformed_trials).

Reply:
Good catch!

Reply:
Fixed in aa72f1f9b7c21c97ad466444fa987bea1ff060d2